src/mem/address_space.h

```diff
@@ -0,0 +1,204 @@
+#include "../pal/pal.h"
+#include "ds/address.h"
+#include "ds/flaglock.h"
+
+#include <array>
+#include <iostream>
+namespace snmalloc
+{
+  template<typename Pal>
+  class AddressSpaceManager : Pal
+  {
+    /**
+     * Implements a power of two allocator, where all blocks are aligned to the
+     * same power of two as their size. This is what snmalloc uses to get
+     * alignment of very large sizeclasses.
+     *
+     * It cannot unreserve memory, so this does not require the
+     * usual complexity of a buddy allocator.
+     */
+
+    // There are a maximum of two blocks for any size/align in a range.
+    // One before the point of maximum alignment, and one after.
+    // As we can add multiple ranges the second entry's block may contain
+    // a pointer to subsequent blocks.
```

Comment:
Please make this and the comment below into doc comments and expand them.  It isn't completely clear to me what this is.  I *think* it is an array of doubly linked lists, indexed by the log2(size) of the allocation (though it seems to be larger than it needs to be: if `bits::BITS` is 64, we have at most 51 used entries in this with 4KiB pages, even on systems that use the full 4KiB range).  

Why are you using a size-two array instead of a pair?